File: app/src/main/java/ch/threema/app/emojis/EmojiSpritemapBitmap.java
package ch.threema.app.emojis;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import org.slf4j.Logger;

import java.io.InputStream;
import java.lang.ref.SoftReference;
import java.util.Objects;

import androidx.annotation.AnyThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import static ch.threema.base.utils.LoggingKt.getThreemaLogger;

public class EmojiSpritemapBitmap {
    private static final Logger logger = getThreemaLogger("EmojiSpritemapBitmap");

    private final Context context;
    private final EmojiGroup emojiGroup;
    private final int spritemapId;
    private final int spritemapInSampleSize;

    private SoftReference<Bitmap> bitmapReference;

    public EmojiSpritemapBitmap(@NonNull Context context, EmojiGroup emojiGroup, int spritemapId, int spritemapInSampleSize) {
        this.context = context.getApplicationContext();
        this.emojiGroup = emojiGroup;
        this.spritemapId = spritemapId;
        this.spritemapInSampleSize = spritemapInSampleSize;
    }

    @Nullable
    @AnyThread
    public Bitmap getSpritemapBitmap() {
        return isSpritemapLoaded() ? bitmapReference.get() : null;
    }

    @AnyThread
    public boolean isSpritemapLoaded() {
        return bitmapReference != null && bitmapReference.get() != null;
    }

    @WorkerThread
    public Bitmap loadSpritemapAsset() {
        try {
            if (emojiGroup.getAssetPath(this.spritemapId) != null) {
                logger.debug("*** Loading Emoji spritemap for group " + this.emojiGroup.getAssetPath(this.spritemapId));

                Bitmap spritemapBitmap = null;
                try (InputStream is = context.getAssets().open(emojiGroup.getAssetPath(this.spritemapId))) {
                    BitmapFactory.Options opts = new BitmapFactory.Options();
                    opts.inJustDecodeBounds = false;
                    opts.inSampleSize = spritemapInSampleSize;
                    spritemapBitmap = BitmapFactory.decodeStream(is, null, opts);
                } catch (Exception e) {
                    logger.error("Could not load emoji bitmap");
                }

                if (spritemapBitmap != null) {
                    bitmapReference = new SoftReference<>(spritemapBitmap);
                    return spritemapBitmap;
                }
            }
        } catch (Exception e) {
            logger.error("Could not find spritemap", e);
        }
        return null;
    }

    @NonNull
    @Override
    public String toString() {
        return Objects.requireNonNull(this.emojiGroup.getAssetPath(this.spritemapId));
    }
}

File: app/src/main/java/ch/threema/app/emojis/EmojiSpritemapBitmap.kt
package ch.threema.app.emojis

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.annotation.AnyThread
import ch.threema.app.utils.DispatcherProvider
import ch.threema.base.utils.getThreemaLogger
import java.lang.ref.SoftReference
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

private val logger = getThreemaLogger("EmojiSpritemapBitmap")

class EmojiSpritemapBitmap(
    private val appContext: Context,
    private val dispatcherProvider: DispatcherProvider,
    private val assetPath: String,
    private val spritemapInSampleSize: Int,
) {
    private val mutex = Mutex()

    private var bitmapReference: SoftReference<Bitmap>? = null

    @get:AnyThread
    var spritemapBitmap: Bitmap?
        get() = bitmapReference?.get()
        private set(value) {
            bitmapReference = value?.let { SoftReference<Bitmap>(value) }
        }

    /**
     * Returns the spritemap bitmap if it is already loaded, or starts loading it and then returns it.
     * It is safe to call this method multiple times, as we ensure internally that the bitmap is at most loaded once
     * and all calls will suspend until that bitmap is loaded.
     * Returns null if loading fails, in which case the method may be called again to retry loading.
     */
    suspend fun getOrLoadSpritemapAsset(): Bitmap? {
        spritemapBitmap?.let {
            return it
        }
        mutex.withLock {
            // we intentionally check spritemapBitmap again here, to avoid loading the bitmap again
            // in case the method gets called multiple times
            spritemapBitmap?.let { bitmap ->
                return bitmap
            }
            try {
                logger.debug("*** Loading Emoji spritemap for group {}", assetPath)
                val spritemapBitmap: Bitmap? = withContext(dispatcherProvider.io) {
                    appContext.assets.open(assetPath).use { inputStream ->
                        val options = BitmapFactory.Options()
                            .apply {
                                inSampleSize = spritemapInSampleSize
                            }
                        BitmapFactory.decodeStream(inputStream, null, options)
                    }
                }

                if (spritemapBitmap != null) {
                    this.spritemapBitmap = spritemapBitmap
                    return spritemapBitmap
                } else {
                    logger.warn("Failed to load spritemap for group {}", assetPath)
                }
            } catch (e: Exception) {
                logger.error("Failed to load emoji spritemap", e)
            }
        }
        return null
    }

    override fun toString() = assetPath
}
